agent/main/lib/WebsocketSession.ts

```diff
@@ -0,0 +1,183 @@
+      // eslint-disable-next-line no-template-curly-in-string
+      .replaceAll('${this.secret}', this.secret)
+      // Use function otherwise replace will try todo some magic
+      .replace('SCRIPT_PLACEHOLDER', () => script);
+
+    const wsScript = `(function ${scriptBody})();`;
+    return wsScript;
+  }
+
+  private proxyScript(): void {
+    const clientId = Math.random();
+    const url = `${this.host}:${this.port}?secret=${this.secret}&clientId=${clientId}`;
+    // This will signal to network manager we are trying to make websocket connection
+    // This is needed later to map clientId to frameId
+    // eslint-disable-next-line no-console
+    fetch(`http://${url}`).catch(error => console.log(error));
+    let callback: WebsocketCallback;
+    try {
+      const socket = new WebSocket(`ws://${url}`);
+      let isReady = false;
+      const queuedCallbacks: { name: string; payload: string }[] = [];
+
+      const sendOverSocket = (name: string, payload: string): void => {
+        try {
+          socket.send(JSON.stringify({ name, payload }));
+        } catch (error) {
+          // eslint-disable-next-line no-console
+          console.log(`failed to send over websocket: ${error}`);
+        }
+      };
+
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      callback = (name, payload): void => {
+        if (!isReady) {
+          queuedCallbacks.push({ name, payload });
+          return;
+        }
+        sendOverSocket(name, payload);
+      };
+
+      socket.addEventListener('open', _event => {
+        let queuedCallback = queuedCallbacks.shift();
+        while (queuedCallback) {
+          sendOverSocket(queuedCallback.name, queuedCallback.payload);
+          queuedCallback = queuedCallbacks.shift();
+        }
+        // Only ready when all older messages have been send so we
+        // keep original order of messages.
+        isReady = true;
+      });
+    } catch (error) {
+      // eslint-disable-next-line no-console
+      console.log(`failed to use websocket: ${error}`);
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-unused-expressions
+    SCRIPT_PLACEHOLDER;
+  }
+
+  private handleUpgrade(request: IncomingMessage, socket: Socket, head: Buffer): void {
+    const url = new URL(request.url, `ws://${this.host}`);
+    // Close and dont send 403 so this acts as an invisible websocket server
+    if (url.searchParams.get('secret') !== this.secret) {
+      socket.destroy();
+    }
+
+    const clientId = url.searchParams.get('clientId');
+    this.wss.handleUpgrade(request, socket as Socket, head, ws => {
+      this.wss.emit('connection', ws, request, clientId);
+    });
+  }
+
+  private handleConnection(ws: Websocket, request: IncomingMessage, clientId: string): void {
+    ws.on('error', error => log.error('WebsocketSession.ConnectionError', { error }));
+    ws.on('message', this.handleMessage.bind(this, clientId));
```

Comment:
I can't find anywhere where we're expecting to handle disconnected web sockets -like on navigations. Will it just work because of the frames? Or are we going to have a timing issue with the time between disconnecting on one page, and reconnecting on the next (in case of mainframe), or in case of a navigation inside a frame

Reply:
Added commit to cleanup. Only problem was small memory leak. This new approach doesnt have all the issues the old context id system had.